R/comment_linters.R

```diff
@@ -0,0 +1,169 @@
+ops <- list(
+  "+",
+  #"-",
+  "=",
+  "==",
+  "!=",
+  "<=",
+  ">=",
+  "<-",
+  "<<-",
+  "<",
+  ">",
+  "->",
+  "->>",
+  "%%",
+  "/",
+  "^",
+  "*",
+  "**",
+  "|",
+  "||",
+  "&",
+  "&&",
+  rex("%", except_any_of("%"), "%"))
+
+#' @describeIn linters Check that there is no commented code outside roxygen
+#' blocks
+#' @export
+commented_code_linter <- function(source_file) {
+  res <- re_matches(source_file$file_lines,
+                    rex("#", any_spaces,
+                        capture(name = "code",
+                          # except("'"),
+                          anything,
+                          or(some_of("{}[]"), # code-like parentheses
+                            or(ops), # any operator
+                            group(graphs, "(", anything, ")"), # a function call
+                            group("!", alphas) # a negation
+                            ),
+                          anything
+                        )
+                    ),
+                    global = FALSE, locations = TRUE)
+
+  line_numbers <- rownames(na.omit(res))
+  lapply(line_numbers, function(line_number) {
+    line <- source_file$file_lines[as.numeric(line_number)]
+    is_parsable <- parsable(substr(line,
+                                   res[line_number, "code.start"],
+                                   res[line_number, "code.end"]))
+    if (is_parsable) {
+      Lint(
+        filename = source_file$filename,
+        line_number = line_number,
+        column_number = res[line_number, "code.start"],
+        type = "style",
+        message = "Commented code should be removed.",
+        line = line,
+        linter = "commented_code_linter",
+        ranges = list(c(res[line_number, "code.start"], res[line_number, "code.end"]))
+        )
+    }
+  })
+}
+
+# is given text parsable
+parsable <- function(x) {
+  if (is.null(x)) {
+    return(FALSE)
+  }
+  res <- try(parse(text = x), silent = TRUE)
+
+  !inherits(res, "try-error")
+}
+
+
+
+#' @describeIn linters  Check that the source contains no TODO comments (case-insensitive).
+#' @param todo  Vector of strings that identify TODO comments.
+#' @export
+todo_comment_linter <- function(todo=c("todo", "fixme")) {
+  local({
+    regex <- rex(capture(name="match", one_or_more("#"), any_spaces, or(todo)))
+    code_locs <- integer()
+    function(source_file) {
+      file_lines <- source_file[["file_lines"]]
+      if (is.null(file_lines)) {
+        # code section: register where expression ends on the line
+        code_locs <<- register_code_locations(source_file, code_locs)
+      } else {
+        # non-code section: detect and report TODOs
+        if (length(file_lines) == 0L) {file_lines <- ""}  # workaround rex bug goo.gl/2rbZDy
```

Comment:
I would rather have a full link rather than short url, as the full link contains information that can be used to retrieve the issue even if the link breaks in the future, such as repo username and issue number.